### PR TITLE
Export-on-demand

### PR DESCRIPTION
**Pull request by micheles**
Now all outputs are stored in the datastore and can be exported on demand. See https://bugs.launchpad.net/oq-engine/+bug/1452363.
The tests are running here: https://ci.openquake.org/job/zdevel_oq-risklib/619/

PS: also closes https://bugs.launchpad.net/oq-engine/+bug/1448580, https://bugs.launchpad.net/oq-engine/+bug/1448584

### CODE REVIEW

File: openquake/commonlib/export/hazard.py [hunk 1]
```diff
@@ -61,53 +64,54 @@ def __iter__(self):
...
-        rows.append([sesrup.tag, sesrup.seed])
-    save_csv(dest, sorted(rows, key=operator.itemgetter(0)))
-    return dest
 
 
-@ds_export.add(('sitecol', 'csv'))
+@export.add(('sitecol', 'csv'))
 def export_sitecol_csv(key, dstore):
-    dest = dstore.export_path(key)
+    dest = dstore.export_path(*key)
```
Comment (ptormene) [inline openquake/commonlib/export/hazard.py]: Please check this

File: openquake/commonlib/export/hazard.py [hunk 2]
```diff
@@ -289,7 +302,101 @@ def export_hazard_curves_csv(key, export_dir, fname, sitecol, curves_by_imt,
     return {fname: dest}
 
 
-@export.add(('hazard_curves', 'xml'))
+def hazard_curve_name(ekey, kind, rlzs_assoc, samples):
```
Comment (ptormene) [inline openquake/commonlib/export/hazard.py]: Please rename samples -> sampling

File: openquake/commonlib/calculators/event_based_risk.py [hunk 3]
```diff
@@ -61,29 +56,49 @@ def event_loss(riskinputs, riskmodel, rlzs_assoc, monitor):
...
+                ad = AccumDict(
+                    data=data, loss=sum(losses), ins_loss=sum(ins_losses),
+                    nonzero=sum(1 for loss in losses if loss),
+                    total=len(losses))
+                try:
+                    a = acc_rlz[out.loss_type, tag]
+                except KeyError:
+                    acc_rlz[out.loss_type, tag] = ad
+                else:
+                    a += ad
```
Comment (ptormene) [inline openquake/commonlib/calculators/event_based_risk.py]: The above names could be made more explicit (a, ad, acc, acc_rlz, ecc.), or some further comments might be added

File: openquake/commonlib/export/hazard.py [hunk 4]
```diff
@@ -61,53 +64,58 @@ def __iter__(self):
...
 
-@export.add(('ses', 'xml'))
-def export_ses_xml(key, export_dir, fname, ses_coll):
-    """
-    Export a Stochastic Event Set Collection
-    """
-    dest = os.path.join(export_dir, fname)
+@export.add(('sescollection', 'xml'), ('sescollection', 'csv'))
+def export_ses_xml(ekey, dstore):
+    fmt = ekey[-1]
```
Comment (ptormene) [inline openquake/commonlib/export/hazard.py]: you might use here the same docstring as in `export_sitecol_csv`

File: openquake/commonlib/export/hazard.py [hunk 5]
```diff
@@ -289,7 +306,101 @@ def export_hazard_curves_csv(key, export_dir, fname, sitecol, curves_by_imt,
...
+            export_uhs_csv(ekey, dstore.export_dir, fname, sitecol, hcurves)
+        else:
+            export_hazard_curves_csv(ekey, dstore.export_dir, fname, sitecol,
+                                     hcurves, oq.imtls)
+    return fnames
+
+
+@export.add(('gmf_by_trt_gsim', 'xml'), ('gmf_by_trt_gsim', 'csv'))
+def export_gmf(ekey, dstore):
+    sitecol = dstore['sitecol']
```
Comment (ptormene) [inline openquake/commonlib/export/hazard.py]: Same as above

File: openquake/commonlib/export/hazard.py [hunk 6]
```diff
@@ -289,7 +306,101 @@ def export_hazard_curves_csv(key, export_dir, fname, sitecol, curves_by_imt,
...
+        fname = '%s-smltp_%s-gsimltp_%s%s.%s' % (
+            prefix, smlt_path, rlz.gsim_rlz.uid, suffix, fmt)
+    else:  # GSIM logic tree realization used in scenario calculators
+        fname = '%s_%s.%s' % (prefix, rlz.uid, fmt)
+    return fname
+
+
+@export.add(('/hcurves', 'csv'), ('/hmaps', 'csv'), ('/uhs', 'csv'))
+def export_hcurves_csv(ekey, dstore):
+    """
```
Comment (ptormene) [inline openquake/commonlib/export/hazard.py]: Same as above

File: openquake/commonlib/export/risk.py [hunk 7]
```diff
@@ -19,15 +19,188 @@
...
+         for asset in assets], asset_dt)
+    return asset_data
+
+
+# ############################### exporters ############################## #
+
+# this is used by classical_risk from csv
+@export.add(('/avg_losses', 'csv'))
+def export_avg_losses(ekey, dstore):
+    avg_losses = dstore[ekey[0] + '/rlzs']
```
Comment (ptormene) [inline openquake/commonlib/export/risk.py]: Same as above

File: openquake/commonlib/export/risk.py [hunk 8]
```diff
@@ -19,15 +19,188 @@
...
+        data = compose_arrays(assets, losses)
+        writers.write_csv(dest, data, fmt='%10.6E', header=columns)
+        fnames.append(dest)
+    return fnames
+
+
+@export.add(('/loss_curves', 'csv'), ('/loss_maps', 'csv'),
+            ('/agg_loss_curve', 'csv'))
+def export_loss_curves(ekey, dstore):
+    name = ekey[0][1:]
```
Comment (ptormene) [inline openquake/commonlib/export/risk.py]: Same as above

File: openquake/commonlib/export/risk.py [hunk 9]
```diff
@@ -19,15 +19,188 @@
...
+            data = compose_arrays(assets, curves[loss_type])
+        dest = os.path.join(
+            export_dir, '%s-%s-%s.csv' % (prefix, loss_type, name))
+        writers.write_csv(dest, data, fmt='%10.6E', header=columns)
+        fnames.append(dest)
+    return fnames
+
+
+@export.add(('event_loss', 'csv'), ('event_loss_asset', 'csv'))
+def export_event_loss(ekey, dstore):
```
Comment (ptormene) [inline openquake/commonlib/export/risk.py]: Same as above

File: openquake/commonlib/export/risk.py [hunk 10]
```diff
@@ -19,15 +19,188 @@
...
+                dstore.export_dir, 'rlz-%03d-%s-%s.csv' % (i, loss_type, name))
+            writers.write_csv(dest, sorted(data[loss_type]), fmt='%10.6E')
+            fnames.append(dest)
+    return fnames
+
+
+# TODO: the export is doing too much; probably we should store
+# a better data structure
+@export.add(('damages_by_key', 'xml'))
+def export_damage(ekey, dstore):
```
Comment (ptormene) [inline openquake/commonlib/export/risk.py]: Same as above

File: openquake/commonlib/export/risk.py [hunk 11]
```diff
@@ -46,46 +219,27 @@ def export_dmg_xml(key, export_dir, damage_states, dmg_data, suffix):
     return AccumDict({key: [dest]})
 
 
-@export.add(('asset-loss', 'csv'), ('asset-ins', 'csv'))
-def export_asset_loss_csv(key, export_dir, data, suffix):
-    """
-    Export aggregate losses in CSV.
+@export.add(('damages_by_rlz', 'csv'))
+def export_classical_damage_csv(ekey, dstore):
```
Comment (ptormene) [inline openquake/commonlib/export/risk.py]: Same as above

File: openquake/commonlib/export/risk.py [hunk 12]
```diff
@@ -98,4 +252,66 @@ def export_classical_damage_csv(key, export_dir, fname, damage_states,
...
+
+AggLoss = collections.namedtuple(
+    'AggLoss', 'loss_type unit mean stddev')
+
+PerAssetLoss = collections.namedtuple(  # the loss map
+    'PerAssetLoss', 'loss_type unit asset_ref mean stddev')
+
+
+@export.add(('losses_by_key', 'csv'))
+def export_risk(ekey, dstore):
```
Comment (ptormene) [inline openquake/commonlib/export/risk.py]: Same as above

File: openquake/commonlib/export/risk.py [hunk 13]
```diff
@@ -19,15 +19,188 @@
...
+        f4 = export_dmg_xml(('collapse_map', 'xml'), oqparam.export_dir,
+                            dmg_states, collapse_map, suffix)
+        fnames.extend(sum((f1 + f2 + f3 + f4).values(), []))
+    return sorted(fnames)
 
 
-@export.add(('dmg_dist_per_asset', 'xml'), ('dmg_dist_per_taxonomy', 'xml'),
-            ('dmg_dist_total', 'xml'), ('collapse_map', 'xml'))
 def export_dmg_xml(key, export_dir, damage_states, dmg_data, suffix):
     """
```
Comment (ptormene) [inline openquake/commonlib/export/risk.py]: Please update the docstring

**Comment by ptormene** [comment]
Besides the major comment: "this is a monster", and the minor comments above, LGTM (hopefully)

File: openquake/commonlib/calculators/event_based_risk.py (hunk 3 above)
Comment (micheles) [inline openquake/commonlib/calculators/event_based_risk.py]: Added a TODO comment

File: openquake/commonlib/export/hazard.py (hunk 4 above)
Comment (micheles) [inline openquake/commonlib/export/hazard.py]: Done

File: openquake/commonlib/export/hazard.py (hunk 5 above)
Comment (micheles) [inline openquake/commonlib/export/hazard.py]: Done

File: openquake/commonlib/export/hazard.py (hunk 6 above)
Comment (micheles) [inline openquake/commonlib/export/hazard.py]: Done

File: openquake/commonlib/export/risk.py (hunk 7 above)
Comment (micheles) [inline openquake/commonlib/export/risk.py]: Done

File: openquake/commonlib/export/risk.py (hunk 8 above)
Comment (micheles) [inline openquake/commonlib/export/risk.py]: Done

File: openquake/commonlib/export/risk.py (hunk 9 above)
Comment (micheles) [inline openquake/commonlib/export/risk.py]: Done

File: openquake/commonlib/export/risk.py (hunk 10 above)
Comment (micheles) [inline openquake/commonlib/export/risk.py]: Done

File: openquake/commonlib/export/risk.py (hunk 11 above)
Comment (micheles) [inline openquake/commonlib/export/risk.py]: Done

File: openquake/commonlib/export/risk.py (hunk 12 above)
Comment (micheles) [inline openquake/commonlib/export/risk.py]: Done

File: openquake/commonlib/export/risk.py (hunk 13 above)
Comment (micheles) [inline openquake/commonlib/export/risk.py]: Done

**Pull request merged**